File: src/main/java/editor/util/Stats.java
package editor.util;

import java.math.BigDecimal;
import java.math.BigInteger;

public interface Stats
{

    /**
     * @param n parameter for factorial
     * @return the factorial of n, or n!
     */
    static BigInteger factorial(int n)
    {
        BigInteger f = BigInteger.ONE;
        for (int i = 1; i <= n; i++)
            f = f.multiply(BigInteger.valueOf(i));
        return f;
    }

    /**
     * Calculate n choose k based on the {@link factorial} function.
     *
     * @param n number of items to choose from
     * @param k number of items to choose
     * @return the number of ways to choose k out of n items.
     */
    static BigDecimal nchoosek(int n, int k)
    {
        if (k == 0)
            return BigDecimal.ONE;
        else if (n == 0)
            return BigDecimal.ZERO;
        else
            return new BigDecimal(factorial(n)).divide(new BigDecimal((factorial(n - k).multiply(factorial(k)))));
    }

    /**
     * Calculate a hypergeometric distribution for drawing the given number
     * of cards in a hand of the given size from a deck of the given size,
     * when the given number of cards are successes.
     *
     * @param n number of desired cards
     * @param hand size of hand drawn
     * @param count number of successful cards in deck
     * @param total number of cards in the deck
     * @return the hypergeometric distribution with parameters hand, count, and
     * total and argument n.
     */
    static double hypergeometric(int n, int hand, int count, int total)
    {
        if (hand - (total - count) > n)
            return 0;
        else
            return nchoosek(count, n).multiply(nchoosek(total - count, hand - n)).divide(nchoosek(total, hand)).doubleValue();
    }
    
}

File: src/main/scala/editor/util/Stats.scala
package editor.util

/**
 * Collection of methods to compute some statistical functions.
 * @author Alec Roelke
 */
object Stats {
  /**
   * Compute the factorial of an integer, or the product of all integers from 1 to that number (or 1
   * if the number is 0).
   * 
   * @param n number to compute the factorial of
   * @return n!, or the factorial of n; or 1 if n == 0
   */
  def factorial(n: Int) = {
    if (n < 0)
      throw IllegalArgumentException("negative factorial")
    else if (n == 0)
      BigInt(1)
    else
      (1 to n).map(BigInt(_)).product
  }

  /**
   * Compute "n choose k", or the number of combinations of n items out of a pool of k.
   * Computed by n!/((n-k)!*k!)
   * 
   * @param n number of items in sample
   * @param k number of items in pool
   * @return n choose k
   */
  def nchoosek(n: Int, k: Int) = {
    if (k == 0)
      BigInt(1)
    else if (n == 0)
      BigInt(0)
    else
      n.! / ((n - k).! * k.!)
  }

  /** Add "n choose k" and n! (as n.!) syntax to Int. */
  implicit class IntToNCK(n: Int) {
    def ! = Stats.factorial(n)
    def choose(k: Int) = Stats.nchoosek(n, k)
  }

  /**
   * Compute a hypergeometric distribution, which is the probability of k successes in n draws out of a pool of N items, K of which are successes.
   * 
   * @param n desired number of successes (k)
   * @param hand number of draws (n)
   * @param count number of successes in pool (K)
   * @param total size of pool (N)
   * @return the probability of drawing n desired cards in hand of the given size out of a deck of size total with count desired cards in it
   */
  def hypergeometric(n: Int, hand: Int, count: Int, total: Int) = {
    if (hand - (total - count) > n)
      0.0
    else
      (BigDecimal((count choose n)*((total - count) choose (hand - n)))/BigDecimal((total choose hand))).toDouble
  }
}